File: lib/formdb_nif.ex
# SPDX-License-Identifier: PMPL-1.0-or-later
# FormDB NIF - Elixir wrapper for Rustler NIF
# This module delegates to the Erlang formdb_nif module

defmodule FormdbNif do
  @moduledoc """
  Elixir wrapper for FormDB Rustler NIF.
  Provides low-level access to FormDB operations.
  Delegates to the Erlang :formdb_nif module.
  """

  # Ensure the Erlang module is compiled first
  # The actual NIF loading happens in the Erlang module

  @doc "Get FormDB version as {major, minor, patch}"
  def version, do: :formdb_nif.version()

  @doc "Open a database connection"
  def db_open(path), do: :formdb_nif.db_open(path)

  @doc "Close a database connection"
  def db_close(db_handle), do: :formdb_nif.db_close(db_handle)

  @doc "Begin a transaction (mode: 'read_only' or 'read_write')"
  def txn_begin(db_handle, mode), do: :formdb_nif.txn_begin(db_handle, mode)

  @doc "Commit a transaction"
  def txn_commit(txn_handle), do: :formdb_nif.txn_commit(txn_handle)

  @doc "Abort a transaction"
  def txn_abort(txn_handle), do: :formdb_nif.txn_abort(txn_handle)

  @doc "Apply a CBOR operation within a transaction"
  def apply(txn_handle, cbor_binary), do: :formdb_nif.apply(txn_handle, cbor_binary)

  @doc "Get database schema (CBOR-encoded)"
  def schema(db_handle), do: :formdb_nif.schema(db_handle)

  @doc "Get journal entries since a sequence number (CBOR-encoded)"
  def journal(db_handle, since), do: :formdb_nif.journal(db_handle, since)
end

File: src/formdb_nif.erl
% SPDX-License-Identifier: PMPL-1.0-or-later
% FormDB NIF - Erlang module loader for Rustler

-module(formdb_nif).

-export([version/0, db_open/1, db_close/1, txn_begin/2, txn_commit/1, txn_abort/1, apply/2, schema/1, journal/2]).

-on_load(init/0).

init() ->
    SoName = case code:priv_dir(formdb_http) of
        {error, bad_name} ->
            case filelib:is_dir(filename:join(["..", priv])) of
                true ->
                    filename:join(["..", priv, native, libformdb_nif]);
                _ ->
                    filename:join([priv, native, libformdb_nif])
            end;
        Dir ->
            filename:join([Dir, native, libformdb_nif])
    end,
    erlang:load_nif(SoName, 0).

%% Stub functions (replaced by NIF)
version() ->
    erlang:nif_error(nif_not_loaded).

db_open(_Path) ->
    erlang:nif_error(nif_not_loaded).

db_close(_Handle) ->
    erlang:nif_error(nif_not_loaded).

txn_begin(_Handle, _Mode) ->
    erlang:nif_error(nif_not_loaded).

txn_commit(_Txn) ->
    erlang:nif_error(nif_not_loaded).

txn_abort(_Txn) ->
    erlang:nif_error(nif_not_loaded).

apply(_Txn, _CborBinary) ->
    erlang:nif_error(nif_not_loaded).

schema(_Handle) ->
    erlang:nif_error(nif_not_loaded).

journal(_Handle, _Since) ->
    erlang:nif_error(nif_not_loaded).
